ramp-database/ramp_database/tools/contributivity.py

```diff
@@ -0,0 +1,338 @@
+import logging
+import numpy as np
+
+from rampwf.utils import get_score_cv_bags
+
+from ..model import CVFold
+from ..model import SubmissionSimilarity
+
+from ._query import select_event_by_name
+from ._query import select_submissions_by_state
+
+logger = logging.getLogger('RAMP-DATABASE')
+
+
+def compute_historical_contributivity(session, event_name):
+    """Compute historical contributivities of an event using
+       contributivities from blending and credits.
+
+    Parameters
+    ----------
+    session : :class:`sqlalchemy.orm.Session`
+        The session to directly perform the operation on the database.
+    event_name : str
+        The event associated to the submission.
+    """
+    submissions = select_submissions_by_state(
+        session, event_name, state='scored')
+    submissions.sort(key=lambda x: x.submission_timestamp, reverse=True)
+    for submission in submissions:
+        submission.historical_contributivity = 0.0
+    for submission in submissions:
+        submission.historical_contributivity += submission.contributivity
+        submission_similaritys = session.query(SubmissionSimilarity).filter_by(
+            type='target_credit', target_submission=submission).all()
+        if submission_similaritys:
+            # if a target team enters several credits to a source submission
+            # we only take the latest
+            submission_similaritys.sort(
+                key=lambda x: x.timestamp, reverse=True)
+            processed_submissions = []
+            historical_contributivity = submission.historical_contributivity
+            for submission_similarity in submission_similaritys:
+                source_submission = submission_similarity.source_submission
+                if source_submission not in processed_submissions:
+                    partial_credit = historical_contributivity *\
+                        submission_similarity.similarity
+                    source_submission.historical_contributivity +=\
+                        partial_credit
+                    submission.historical_contributivity -= partial_credit
+                    processed_submissions.append(source_submission)
+    session.commit()
+
+
+def compute_contributivity(session, event_name):
+    """Blend submissions of an event, compute combined score and
```

Comment:
We should probably have a function which manages the submission blending

Reply:
And the only job here will be to compute the contributivity

Reply:
Not sure what you mean here. It would repeat the same loop (that can be quite heavy sometimes).